# Oakland Race and Equity Main 

This where consolidated cleaning and data exploration for the project will take place.

## Import Libraries

In [1]:
import pandas as pd
pd.set_option('display.max_columns', None)

## Import Data

The data is the ACS (American Community Survey) PUMS (Public Use Microdata) provided by the US census website.
- The data comes as two periods, a 1 year period (2019 latest) and a 5 year (2015-2019 latest)
- Each period has two datasets, person (denotes individuals) and household (denotes households)

2019, 1 year datasets can be found at these links:<br>
<b>Person</b><br>
https://www2.census.gov/programs-surveys/acs/data/pums/2019/1-Year/csv_pca.zip<br>
<b>Household</b><br>
https://www2.census.gov/programs-surveys/acs/data/pums/2019/1-Year/csv_hca.zip

For readability, we renamed 1 year csv files to to `2019pCA1.csv` and `2019hCA1.csv` respectively.

In [2]:
pca191 = pd.read_csv("./Data/2019pCA1.csv")
hca191 = pd.read_csv("./Data/2019hCA1.csv")

### Minor data exploration

In [3]:
pca191.head()

RT       SERIALNO  DIVISION  SPORDER  PUMA  REGION  ST   ADJINC  PWGTP  \
0  P  2019GQ0000003         9        1  3704       4   6  1010145     21   
1  P  2019GQ0000009         9        1  7322       4   6  1010145     34   
2  P  2019GQ0000013         9        1  5904       4   6  1010145     28   
3  P  2019GQ0000023         9        1  7107       4   6  1010145    127   
4  P  2019GQ0000024         9        1  8900       4   6  1010145    103   

   AGEP  CIT  CITWP  COW  DDRS  DEAR  DEYE  DOUT  DPHY  DRAT  DRATX  DREM  \
0    58    1    NaN  1.0   1.0     1     2   1.0   1.0   NaN    NaN   1.0   
1    66    1    NaN  NaN   2.0     2     2   1.0   1.0   NaN    NaN   2.0   
2    18    1    NaN  1.0   2.0     2     2   2.0   2.0   NaN    NaN   2.0   
3    58    1    NaN  NaN   2.0     2     2   2.0   2.0   NaN    2.0   2.0   
4    18    1    NaN  1.0   2.0     2     2   2.0   2.0   NaN    NaN   2.0   

   ENG  FER  GCL  GCM  GCR  HIMRKS  HINS1  HINS2  HINS3  HINS4  HINS5  HINS6  \
0  NaN  NaN  2.0  NaN  NaN       0      2      2      1      2      2      2   
1  NaN  NaN  2.0  NaN  NaN       0      2      2      2      2      2      2   
2  NaN  2.0  NaN  NaN  NaN       0      2      2      2      2      2      2   
3  NaN  NaN  2.0  NaN  NaN       0      2      2      2      2      2      2   
4  NaN  2.0  NaN  NaN  NaN       1      2      1      2      2      2      2   

   HINS7  INTP  JWMNP  JWRIP  JWTRNS  LANX  MAR  MARHD  MARHM  MARHT  MARHW  \
0      2   0.0    NaN    NaN     NaN   2.0    5    NaN    NaN    NaN    NaN   
1      2   0.0    NaN    NaN     NaN   2.0    3    2.0    2.0    3.0    2.0   
2      2   0.0    NaN    NaN     NaN   2.0    5    NaN    NaN    NaN    NaN   
3      2   0.0    NaN    NaN     NaN   2.0    5    NaN    NaN    NaN    NaN   
4      2   0.0    NaN    NaN     NaN   2.0    5    NaN    NaN    NaN    NaN   

   MARHYP  MIG  MIL  MLPA  MLPB  MLPCD  MLPE  MLPFG  MLPH  MLPI  MLPJ  MLPK  \
0     NaN  1.0  4.0   NaN   NaN    NaN   NaN    NaN   NaN   NaN   NaN   NaN   
1  1994.0  1.0  4.0   NaN   NaN    NaN   NaN    NaN   NaN   NaN   NaN   NaN   
2     NaN  3.0  4.0   NaN   NaN    NaN   NaN    NaN   NaN   NaN   NaN   NaN   
3     NaN  1.0  3.0   NaN   NaN    NaN   NaN    NaN   NaN   NaN   NaN   NaN   
4     NaN  3.0  4.0   NaN   NaN    NaN   NaN    NaN   NaN   NaN   NaN   NaN   

   NWAB  NWAV  NWLA  NWLK  NWRE  OIP  PAP  RELSHIPP  RETP  SCH  SCHG  SCHL  \
0   2.0   2.0   2.0   1.0   3.0  0.0  0.0        38   0.0  1.0   NaN  18.0   
1   3.0   5.0   3.0   3.0   3.0  0.0  0.0        37   0.0  1.0   NaN  16.0   
2   2.0   5.0   2.0   2.0   3.0  0.0  0.0        38   0.0  2.0  15.0  16.0   
3   3.0   5.0   3.0   3.0   3.0  0.0  0.0        37   0.0  1.0   NaN  13.0   
4   2.0   3.0   2.0   2.0   2.0  0.0  0.0        38   0.0  3.0  15.0  16.0   

   SEMP  SEX  SSIP  SSP     WAGP  WKHP  WKL  WKWN  WRK  YOEP  ANC  ANC1P  \
0   0.0    2   0.0  0.0  23100.0  20.0  1.0  32.0  2.0   NaN    3    995   
1   0.0    1   0.0  0.0      0.0   NaN  3.0   NaN  NaN   NaN    1     51   
2   0.0    2   0.0  0.0   6000.0  15.0  1.0  32.0  2.0   NaN    1     50   
3   0.0    1   0.0  0.0      0.0   NaN  3.0   NaN  NaN   NaN    4    999   
4   0.0    2   0.0  0.0   2000.0  35.0  1.0   6.0  2.0   NaN    2    931   

   ANC2P  DECADE  DIS  DRIVESP  ESP  ESR  FOD1P  FOD2P  HICOV  HISP    INDP  \
0    999     NaN    1      NaN  NaN  3.0    NaN    NaN      1     1  6190.0   
1    999     NaN    1      NaN  NaN  6.0    NaN    NaN      2     1     NaN   
2    999     NaN    2      NaN  NaN  6.0    NaN    NaN      2     1  8680.0   
3    999     NaN    2      NaN  NaN  6.0    NaN    NaN      2     1     NaN   
4     88     NaN    2      NaN  NaN  6.0    NaN    NaN      1     1  8670.0   

   JWAP  JWDP  LANP  MIGPUMA  MIGSP  MSP NAICSP  NATIVITY  NOP  OC    OCCP  \
0   NaN   NaN   NaN      NaN    NaN  6.0   4853         1  NaN NaN  9142.0   
1   NaN   NaN   NaN      NaN    NaN  4.0    NaN         1  NaN NaN     NaN   
2   NaN   NaN   

In [4]:
hca191.head()

RT       SERIALNO  DIVISION  PUMA  REGION  ST   ADJHSG   ADJINC  WGTP  NP  \
0  H  2019GQ0000003         9  3704       4   6  1000000  1010145     0   1   
1  H  2019GQ0000009         9  7322       4   6  1000000  1010145     0   1   
2  H  2019GQ0000013         9  5904       4   6  1000000  1010145     0   1   
3  H  2019GQ0000023         9  7107       4   6  1000000  1010145     0   1   
4  H  2019GQ0000024         9  8900       4   6  1000000  1010145     0   1   

   TYPE  ACCESS  ACR  AGS  BATH  BDSP  BLD  BROADBND  COMPOTHX  CONP  DIALUP  \
0     3     NaN  NaN  NaN   NaN   NaN  NaN       NaN       NaN   NaN     NaN   
1     2     NaN  NaN  NaN   NaN   NaN  NaN       NaN       NaN   NaN     NaN   
2     3     NaN  NaN  NaN   NaN   NaN  NaN       NaN       NaN   NaN     NaN   
3     2     NaN  NaN  NaN   NaN   NaN  NaN       NaN       NaN   NaN     NaN   
4     3     NaN  NaN  NaN   NaN   NaN  NaN       NaN       NaN   NaN     NaN   

   ELEFP  ELEP   FS  FULFP  FULP  GASFP  GASP  HFL  HISPEED  HOTWAT  INSP  \
0    NaN   NaN  1.0    NaN   NaN    NaN   NaN  NaN      NaN     NaN   NaN   
1    NaN   NaN  2.0    NaN   NaN    NaN   NaN  NaN      NaN     NaN   NaN   
2    NaN   NaN  2.0    NaN   NaN    NaN   NaN  NaN      NaN     NaN   NaN   
3    NaN   NaN  2.0    NaN   NaN    NaN   NaN  NaN      NaN     NaN   NaN   
4    NaN   NaN  2.0    NaN   NaN    NaN   NaN  NaN      NaN     NaN   NaN   

   LAPTOP  MHP  MRGI  MRGP  MRGT  MRGX  OTHSVCEX  REFR  RMSP  RNTM  RNTP  \
0     NaN  NaN   NaN   NaN   NaN   NaN       NaN   NaN   NaN   NaN   NaN   
1     NaN  NaN   NaN   NaN   NaN   NaN       NaN   NaN   NaN   NaN   NaN   
2     NaN  NaN   NaN   NaN   NaN   NaN       NaN   NaN   NaN   NaN   NaN   
3     NaN  NaN   NaN   NaN   NaN   NaN       NaN   NaN   NaN   NaN   NaN   
4     NaN  NaN   NaN   NaN   NaN   NaN       NaN   NaN   NaN   NaN   NaN   

   RWAT  RWATPR  SATELLITE  SINK  SMARTPHONE  SMP  STOV  TABLET  TEL  TEN  \
0   NaN     NaN        NaN   NaN         NaN  NaN   NaN     NaN  NaN  NaN   
1   NaN     NaN        NaN   NaN         NaN  NaN   NaN     NaN  NaN  NaN   
2   NaN     NaN        NaN   NaN         NaN  NaN   NaN     NaN  NaN  NaN   
3   NaN     NaN        NaN   NaN         NaN  NaN   NaN     NaN  NaN  NaN   
4   NaN     NaN        NaN   NaN         NaN  NaN   NaN     NaN  NaN  NaN   

   VACS  VALP  VEH  WATFP  WATP  YBL  CPLT  FES  FINCP  FPARC  GRNTP  GRPIP  \
0   NaN   NaN  NaN    NaN   NaN  NaN   NaN  NaN    NaN    NaN    NaN    NaN   
1   NaN   NaN  NaN    NaN   NaN  NaN   NaN  NaN    NaN    NaN    NaN    NaN   
2   NaN   NaN  NaN    NaN   NaN  NaN   NaN  NaN    NaN    NaN    NaN    NaN   
3   NaN   NaN  NaN    NaN   NaN  NaN   NaN  NaN    NaN    NaN    NaN    NaN   
4   NaN   NaN  NaN    NaN   NaN  NaN   NaN  NaN    NaN    NaN    NaN    NaN   

   HHL  HHLANP  HHT  HHT2  HINCP  HUGCL  HUPAC  HUPAOC  HUPARC  KIT  LNGI  \
0  NaN     NaN  NaN   NaN    NaN    NaN    NaN     NaN     NaN  NaN   NaN   
1  NaN     NaN  NaN   NaN    NaN    NaN    NaN     NaN     NaN  NaN   NaN   
2  NaN     NaN  NaN   NaN    NaN    NaN    NaN     NaN     NaN  NaN   NaN   
3  NaN     NaN  NaN   NaN    NaN    NaN    NaN     NaN     NaN  NaN   NaN   
4  NaN     NaN  NaN   NaN    NaN    NaN    NaN     NaN     NaN  NaN   NaN   

   MULTG  MV  NOC  NPF  NPP  NR  NRC  OCPIP  PARTNER  PLM  PLMPRP  PSF  R18  \
0    NaN NaN  NaN  NaN  NaN NaN  NaN    NaN      NaN  NaN     NaN  NaN  NaN   
1    NaN NaN  NaN  NaN  NaN NaN  NaN    NaN      NaN  NaN     NaN  NaN  NaN   
2    NaN NaN  NaN  NaN  NaN NaN  NaN    NaN      NaN  NaN     NaN  NaN  NaN   
3    NaN NaN  NaN  NaN  NaN NaN  NaN    NaN      NaN  NaN     NaN  NaN  NaN   
4    NaN NaN  NaN  NaN  NaN NaN  NaN    NaN      NaN  NaN     NaN  NaN  NaN   

   R60  R65  RESMODE  SMOCP  SMX  SRNT  SVAL  TAXAMT  WIF  WKEXREL  WORKSTAT  \
0  NaN  NaN      NaN    NaN  NaN   NaN   NaN     NaN  NaN      NaN       NaN   
1  NaN  NaN      NaN    NaN  NaN   NaN   NaN     NaN  NaN      NaN       NaN   
2  NaN  Na

In [5]:
print("California House data dimensions: " + str(hca191.shape))
print("California Person data dimensions: " + str(pca191.shape))

California House data dimensions: (159756, 235)
California Person data dimensions: (380091, 288)


## Cleaning

### Filtering Relevant Columns

There are 288 columns in the population dataset, and only about 20 indicators we need to calculate using this data, meaning most are useless for the purpose of this project. We will cut out the columns that are not used in indicator calculations and are not useful to us in analysis.

In [6]:
pca191 = pca191.loc[:, ["SERIALNO", "SPORDER", "PUMA", "PWGTP", "AGEP", "CIT", "COW", "ENG", "FER", "JWMNP"
               , "JWTRNS", "MAR", "MIL", "SCH", "SCHL", "RELSHIPP", "SEX", "PAP", "INTP", "SSIP", "SSP", "WAGP"
               , "OIP", "RETP", "SEMP", "PERNP", "PINCP", "WKL", "DIS", "ESR", "HICOV", "HISP"
               , "PAOC", "POVPIP", "RAC1P", "RACASN", "RACBLK", "RACWHT", "RACSOR", "SCIENGP", "WKWN", "WKHP"
               , "SOCP", "RELSHIPP"]].copy()

In [7]:
hca191 = hca191.loc[:, ["SERIALNO", "PUMA", "NP", "ACCESS", "ACR", "BATH", "FS", "ELEP", "FULP", "GASP", "HISPEED"
              , "LAPTOP", "RNTP", "RWATPR", "TEN", "VALP", "VEH", "WATP", "HINCP", "HUPAC", "KIT", "PLM"
              , "GRPIP", "RMSP"]].copy()

In [8]:
print("New P dataset dimensions: " + str(pca191.shape))

New P dataset dimensions: (380091, 44)


In [9]:
print("New H dataset dimensions: " + str(hca191.shape))

New H dataset dimensions: (159756, 24)


### Filtering PUMA codes

We don't want to use all of California's data for the main project, just Oakland data. ~However, we will make a separate dataset for the sorrounding areas for plotting later.~ (Still unsure of wether I'll do this)

Oakland uses PUMA codes codes 102-105, so we take out data not in these areas.

We will make a copy of the orignal data to make the subsets, to avoid a `SettingwithCopyWarning` error and we may use the entire dataset later, so we don't want to overwrite it.

In [10]:
pca191_Ok = pca191.loc[(pca191["PUMA"] > 101) & (pca191["PUMA"] <= 105)].copy()
hca191_Ok = hca191.loc[(hca191["PUMA"] > 101) & (hca191["PUMA"] <= 105)].copy()

In [11]:
print("Oakland House data dimensions: " + str(pca191_Ok.shape))
print("Oakland Population data dimensions: " + str(hca191_Ok.shape))

Oakland House data dimensions: (6147, 44)
Oakland Population data dimensions: (2831, 24)


### Making a 'RACE' column

For the purpose of this report, we are only examining 5 races
- African American
- Asian
- Latino
- White
- Other

The races are denoted by the `RAC1P` column in the population dataset. They're coded as follows:

- African American - 2
- Asian - 6
- White - 1
- Other - 3, 4, 5, 7, 8, 9

Latinos can be any race, so the `HISP` column denotes people who identify as Latino instead of `RAC1P`. Non-Latinos must have a value of <i>1</i> in the `HISP` column to not be counted as hispanic. Otherwise, if `HISP` is any value other than <i>1</i>, the person is counted as Latino regardless of race.

In [12]:
pca191_Ok.loc[(pca191_Ok.RAC1P == 1) & (pca191_Ok.HISP == 1), "RACE"] = "White"
pca191_Ok.loc[(pca191_Ok.RAC1P == 2) & (pca191_Ok.HISP == 1), "RACE"] = "African American"
pca191_Ok.loc[(pca191_Ok.RAC1P == 6) & (pca191_Ok.HISP == 1), "RACE"] = "Asian"
pca191_Ok.loc[(pca191_Ok.RAC1P.isin([3, 4, 5, 7, 8, 9])) & (pca191_Ok.HISP == 1), "RACE"] = "Other"
pca191_Ok.loc[pca191_Ok.HISP != 1, "RACE"] = "Latino"

We will check the logic with a random sample.

In [13]:
pca191_Ok.sample(n=5, random_state=8)[["RAC1P", "HISP", "RACE"]]

RAC1P  HISP              RACE
212339      6     1             Asian
155667      1     1             White
20131       3     1             Other
372887      2     1  African American
252313      8    11            Latino

The races have been correctly assigned.

## Merging Household data

We will merge household data with a left join, with the person dataset being the left table.

In [14]:
pca191_Ok = pca191_Ok.merge(hca191_Ok, how = 'left', on = 'SERIALNO')

In [15]:
pca191_Ok.tail(5)

SERIALNO  SPORDER  PUMA_x  PWGTP  AGEP  CIT  COW  ENG  FER  JWMNP  \
6142  2019HU1409863        1     103     53    38    1  4.0  NaN  NaN   30.0   
6143  2019HU1409863        2     103     46    37    3  2.0  1.0  2.0   30.0   
6144  2019HU1409863        3     103     62     2    1  NaN  NaN  NaN    NaN   
6145  2019HU1412346        1     103     98    36    1  2.0  NaN  2.0   25.0   
6146  2019HU1412346        2     103    110    38    1  3.0  NaN  NaN   20.0   

      JWTRNS  MAR  MIL  SCH  SCHL  RELSHIPP  SEX     PAP    INTP  SSIP  SSP  \
6142     1.0    1  4.0  1.0  24.0        20    1     0.0  2000.0   0.0  0.0   
6143     2.0    1  4.0  1.0  23.0        21    2  4000.0     0.0   0.0  0.0   
6144     NaN    5  NaN  NaN   NaN        25    1     NaN     NaN   NaN  NaN   
6145     1.0    1  4.0  1.0  22.0        20    2     0.0     0.0   0.0  0.0   
6146     1.0    1  4.0  1.0  22.0        21    1     0.0     0.0   0.0  0.0   

         WAGP  OIP  RETP     SEMP    PERNP    PINCP  WKL  DIS  ESR  HICOV  \
6142  73000.0  0.0   0.0      0.0  73000.0  75000.0  1.0    2  1.0      1   
6143  89000.0  0.0   0.0      0.0  89000.0  93000.0  1.0    2  1.0      1   
6144      NaN  NaN   NaN      NaN      NaN      NaN  NaN    2  NaN      1   
6145  82000.0  0.0   0.0      0.0  82000.0  82000.0  1.0    2  1.0      1   
6146  15000.0  0.0   0.0  24000.0  39000.0  39000.0  1.0    2  1.0      1   

      HISP  PAOC  POVPIP  RAC1P  RACASN  RACBLK  RACWHT  RACSOR  SCIENGP  \
6142     1   NaN   501.0      1       0       0       1       0      1.0   
6143     1   1.0   501.0      1       0       0       1       0      1.0   
6144     1   NaN   501.0      1       0       0       1       0      NaN   
6145     1   4.0   501.0      1       0       0       1       0      1.0   
6146     1   NaN   501.0      1       0       0       1       0      2.0   

      WKWN  WKHP    SOCP  RELSHIPP   RACE  PUMA_y  NP  ACCESS  ACR  BATH   FS  \
6142  52.0  50.0  251000        20  White     103   3     1.0  1.0   1.0  2.0   
6143  52.0  60.0  2310XX        21  White     103   3     1.0  1.0   1.0  2.0   
6144   NaN   NaN     NaN        25  White     103   3     1.0  1.0   1.0  2.0   
6145  52.0  40.0  112030        20  White     103   2     1.0  NaN   1.0  2.0   
6146  49.0  25.0  254022        21  White     103   2     1.0  NaN   1.0  2.0   

       ELEP  FULP  GASP  HISPEED  LAPTOP   RNTP  RWATPR  TEN       VALP  VEH  \
6142  420.0   NaN   NaN      1.0     1.0    NaN     9.0  1.0  1200000.0  2.0   
6143  420.0   NaN   NaN      1.0     1.0    NaN     9.0  1.0  1200000.0  2.0   
6144  420.0   NaN   NaN      1.0     1.0    NaN     9.0  1.0  1200000.0  2.0   
6145   70.0   NaN   NaN      1.0     1.0  990.0     9.0  3.0        NaN  2.0   
6146   70.0   NaN   NaN      1.0     1.0  990.0     9.0  3.0        NaN  2.0   

        WATP     HINCP  HUPAC  KIT  PLM  GRPIP  RMSP  
6142  1300.0  168000.0    1.0  1.0  1.0    NaN   5.0  
6143  1300.0  168000.0    1.0  1.0  1.0    NaN   5.0  
6144  1300.0  168000.0    1.0  1.0  1.0    NaN   5.0  
6145     NaN  121000.0    4.0  1.0  1.0   11.0   2.0  
6146     NaN  121000.0    4.0  1.0  1.0   11.0   2.0

In [16]:
pca191_Ok.shape

(6147, 68)

We kept the same number of rows in the persons data and added the household columns appropriately.

## Examine NAs

We'll run a loop to tell me which columns have NA data.

In [17]:
message = "Missing values in the following columns: \n"
for column in pca191_Ok.columns:
    if pca191_Ok[column].isnull().values.any():
        message += column + ", "
print(message)

Missing values in the following columns: 
COW, ENG, FER, JWMNP, JWTRNS, MIL, SCH, SCHL, PAP, INTP, SSIP, SSP, WAGP, OIP, RETP, SEMP, PERNP, PINCP, WKL, ESR, PAOC, POVPIP, SCIENGP, WKWN, WKHP, SOCP, ACCESS, ACR, BATH, ELEP, FULP, GASP, HISPEED, LAPTOP, RNTP, RWATPR, TEN, VALP, VEH, WATP, HINCP, HUPAC, KIT, PLM, GRPIP, RMSP, 


The PUMS dataset classifies NAs as data that is not available because the current user doesn't qualify for the column.

For example, users under 16 have NAs in the `WAGP` column because they can't legally work. Likewise users who haven't gone to college have and NA in the `SCIENGP` column, which denotes a STEM degree.

Because the NA's vary so much in the data, <b>we will not get rid of any NAs until we do statistical analysis and cannot fill them with a <i>0</i> value or the mean as the NAs represent data that cannot exist.</b>

## Duplicate data

The final step of the data cleaning will be to explode the data. Since the PUMS dataset is weighted survey data, each row represents a group of people determined by the `PWGTP` column. For example, the first row in the dataset has a `PWGTP` value of 13, meaning that row accounts for 13 people. 

Because we will be performing statistical and machine learning analysis on the data, we need to take the weights into consideration. A simple way to do this is to duplicate the rows based on the weight column value. For example, going back to the first row, since the `PWGTP` value is 13, we will duplicate the row 13 times. 

In [18]:
pca191_Ok = pca191_Ok.reindex(pca191_Ok.index.repeat(pca191_Ok.PWGTP)).reset_index(drop=True).drop(['PWGTP'], axis=1).rename(columns={"PUMA_x": "PUMA"}).copy()

In [19]:
pca191_Ok.head(5)

SERIALNO  SPORDER  PUMA  AGEP  CIT  COW  ENG  FER  JWMNP  JWTRNS  MAR  \
0  2019GQ0000081        1   102    70    1  NaN  NaN  NaN    NaN     NaN    5   
1  2019GQ0000081        1   102    70    1  NaN  NaN  NaN    NaN     NaN    5   
2  2019GQ0000081        1   102    70    1  NaN  NaN  NaN    NaN     NaN    5   
3  2019GQ0000081        1   102    70    1  NaN  NaN  NaN    NaN     NaN    5   
4  2019GQ0000081        1   102    70    1  NaN  NaN  NaN    NaN     NaN    5   

   MIL  SCH  SCHL  RELSHIPP  SEX  PAP  INTP  SSIP  SSP  WAGP  OIP  RETP  SEMP  \
0  4.0  1.0  16.0        37    2  0.0   0.0   0.0  0.0   0.0  0.0   0.0   0.0   
1  4.0  1.0  16.0        37    2  0.0   0.0   0.0  0.0   0.0  0.0   0.0   0.0   
2  4.0  1.0  16.0        37    2  0.0   0.0   0.0  0.0   0.0  0.0   0.0   0.0   
3  4.0  1.0  16.0        37    2  0.0   0.0   0.0  0.0   0.0  0.0   0.0   0.0   
4  4.0  1.0  16.0        37    2  0.0   0.0   0.0  0.0   0.0  0.0   0.0   0.0   

   PERNP  PINCP  WKL  DIS  ESR  HICOV  HISP  PAOC  POVPIP  RAC1P  RACASN  \
0    0.0    0.0  3.0    1  6.0      1     1   NaN     NaN      2       0   
1    0.0    0.0  3.0    1  6.0      1     1   NaN     NaN      2       0   
2    0.0    0.0  3.0    1  6.0      1     1   NaN     NaN      2       0   
3    0.0    0.0  3.0    1  6.0      1     1   NaN     NaN      2       0   
4    0.0    0.0  3.0    1  6.0      1     1   NaN     NaN      2       0   

   RACBLK  RACWHT  RACSOR  SCIENGP  WKWN  WKHP SOCP  RELSHIPP  \
0       1       0       0      NaN   NaN   NaN  NaN        37   
1       1       0       0      NaN   NaN   NaN  NaN        37   
2       1       0       0      NaN   NaN   NaN  NaN        37   
3       1       0       0      NaN   NaN   NaN  NaN        37   
4       1       0       0      NaN   NaN   NaN  NaN        37   

               RACE  PUMA_y  NP  ACCESS  ACR  BATH   FS  ELEP  FULP  GASP  \
0  African American     102   1     NaN  NaN   NaN  2.0   NaN   NaN   NaN   
1  African American     102   1     NaN  NaN   NaN  2.0   NaN   NaN   NaN   
2  African American     102   1     NaN  NaN   NaN  2.0   NaN   NaN   NaN   
3  African American     102   1     NaN  NaN   NaN  2.0   NaN   NaN   NaN   
4  African American     102   1     NaN  NaN   NaN  2.0   NaN   NaN   NaN   

   HISPEED  LAPTOP  RNTP  RWATPR  TEN  VALP  VEH  WATP  HINCP  HUPAC  KIT  \
0      NaN     NaN   NaN     NaN  NaN   NaN  NaN   NaN    NaN    NaN  NaN   
1      NaN     NaN   NaN     NaN  NaN   NaN  NaN   NaN    NaN    NaN  NaN   
2      NaN     NaN   NaN     NaN  NaN   NaN  NaN   NaN    NaN    NaN  NaN   
3      NaN     NaN   NaN     NaN  NaN   NaN  NaN   NaN    NaN    NaN  NaN   
4      NaN     NaN   NaN     NaN  NaN   NaN  NaN   NaN    NaN    NaN  NaN   

   PLM  GRPIP  RMSP  
0  NaN    NaN   NaN  
1  NaN    NaN   NaN  
2  NaN    NaN   NaN  
3  NaN    NaN   NaN  
4  NaN    NaN   NaN

In [20]:
print("Dimensions for the final person dataset: " + str(pca191_Ok.shape))
print("New DataFrame size: " + str(int(pca191_Ok.memory_usage(index=True).sum()/1000000)) + " MB")

Dimensions for the final person dataset: (625414, 67)
New DataFrame size: 335 MB


## Cleaning 2016 Data

We will proces the 2016 PUMS data in the same way as the 2019 to produce the same type of cleaned file. This will help us validate our results for 2019, by checking that we are calculating the indicators correctly with 2016 data, as we have the results from the 2016 project.


2016, 1 year datasets can be found at these links:<br>
<b>Person (csv_pca.zip)</b><br>
https://www2.census.gov/programs-surveys/acs/data/pums/2016/1-Year/<br>
<b>Household (csv_hca.zip)</b><br>
https://www2.census.gov/programs-surveys/acs/data/pums/2016/1-Year/

In [21]:
pca161 = pd.read_csv("./Data/2016pCA1.csv")
hca161 = pd.read_csv("./Data/2016hCA1.csv")
pca161 = pca161.loc[:, ["SERIALNO", "SPORDER", "PUMA", "PWGTP", "AGEP", "CIT", "COW", "ENG", "FER", "JWMNP"
               , "MAR", "MIL", "SCH", "SCHL", "SEX", "PAP", "INTP", "SSIP", "SSP", "WAGP"
               , "OIP", "RETP", "SEMP", "PERNP", "PINCP", "WKL", "DIS", "ESR", "HICOV", "HISP"
               , "PAOC", "POVPIP", "RAC1P", "RACASN", "RACBLK", "RACWHT", "RACSOR", "SCIENGP", "WKHP"
               , "SOCP", "RELP"]].copy()
hca161 = hca161.loc[:, ["SERIALNO", "PUMA", "NP", "ACCESS", "ACR", "BATH", "FS", "ELEP", "FULP", "GASP", "HISPEED"
              , "LAPTOP", "RNTP", "RWATPR", "TEN", "VALP", "VEH", "WATP", "HINCP", "HUPAC", "KIT", "PLM"
              , "GRPIP", "RMSP"]].copy()
pca161_Ok = pca161.loc[(pca161["PUMA"] > 101) & (pca161["PUMA"] <= 105)].copy()
hca161_Ok = hca161.loc[(hca161["PUMA"] > 101) & (hca161["PUMA"] <= 105)].copy()
pca161_Ok.loc[(pca161_Ok.RAC1P == 1) & (pca161_Ok.HISP == 1), "RACE"] = "White"
pca161_Ok.loc[(pca161_Ok.RAC1P == 2) & (pca161_Ok.HISP == 1), "RACE"] = "African American"
pca161_Ok.loc[(pca161_Ok.RAC1P == 6) & (pca161_Ok.HISP == 1), "RACE"] = "Asian"
pca161_Ok.loc[(pca161_Ok.RAC1P.isin([3, 4, 5, 7, 8, 9])) & (pca161_Ok.HISP == 1), "RACE"] = "Other"
pca161_Ok.loc[pca161_Ok.HISP != 1, "RACE"] = "Latino"
pca161_Ok = pca161_Ok.merge(hca161_Ok, how = 'left', on = 'SERIALNO').copy()
pca161_Ok = pca161_Ok.reindex(pca161_Ok.index.repeat(pca161_Ok.PWGTP)).reset_index(drop=True).drop(['PWGTP'], axis=1).rename(columns={"PUMA_x": "PUMA"}).copy()

In [22]:
print("Dimensions for the final person dataset: " + str(pca161_Ok.shape))
print("New DataFrame size: " + str(int(pca161_Ok.memory_usage(index=True).sum()/1000000)) + " MB")

Dimensions for the final person dataset: (612910, 64)
New DataFrame size: 313 MB


### Test Business Ownership

In [23]:
bo = pca161_Ok.loc[pca161_Ok["COW"] == 7, ["RACE"]].groupby("RACE")['RACE'].count()

In [24]:
ep = pca161_Ok.loc[(pca161_Ok["COW"] != 9) & (pca161_Ok["COW"].notna()), ["RACE"]].groupby("RACE")['RACE'].count()

In [25]:
bo/ep * 100

RACE
African American    1.448461
Asian               2.264658
Latino              2.368998
Other               0.525924
White               3.915227
Name: RACE, dtype: float64

### Home Ownership

In [26]:
dh = pca161_Ok.loc[(pca161_Ok.RELP == 0) & ((pca161_Ok.TEN == 2) | (pca161_Ok.TEN == 1)), ["RACE"]].groupby("RACE")['RACE'].count()

In [27]:
da = pca161_Ok.loc[pca161_Ok.RELP == 0, ["RACE"]].groupby("RACE")['RACE'].count()

In [28]:
1 - dh/da

RACE
African American    0.740715
Asian               0.480818
Latino              0.689713
Other               0.701454
White               0.435802
Name: RACE, dtype: float64

### Energy Cost Burden

In [46]:
pca161_Ok_EC = pca161_Ok.loc[:, ["RELP", "ELEP", "FULP", "GASP", "HINCP", "RACE"]]
pca161_Ok_EC = pca161_Ok_EC.dropna()
pca161_Ok_EC = pca161_Ok_EC.loc[pca161_Ok_EC.RELP == 0]
pca161_Ok_EC.loc[pca161_Ok_EC["ELEP"] == 2, "ELEP"] = 0
pca161_Ok_EC.loc[(pca161_Ok_EC["FULP"] == 1) | (pca161_Ok_EC["FULP"] ==  2), "FULP"] = 0
pca161_Ok_EC.loc[(pca161_Ok_EC["GASP"] == 2) | (pca161_Ok_EC["GASP"] == 3), "GASP"] = 0
pca161_Ok_EC = pca161_Ok_EC.loc[(pca161_Ok_EC["ELEP"] != 1) & (pca161_Ok_EC["GASP"] != 1)]
pca161_Ok_EC = pca161_Ok_EC.loc[pca161_Ok_EC["HINCP"] > 0]

In [43]:
energy = ((pca161_Ok_EC.ELEP * 12) + (pca161_Ok_EC.GASP * 12) + pca161_Ok_EC.FULP)/pca161_Ok_EC.HINCP
pca161_Ok_EC["ENERGY"] = energy

In [44]:
pca161_Ok_EC[["RACE", "ENERGY"]].groupby("RACE").median()

ENERGY
RACE                      
African American  0.023861
Asian             0.013416
Latino            0.018812
Other             0.013935
White             0.010179